File: 2020/day24/day24.py
#!/usr/bin/env python3
# [Day 24: Lobby Layout](https://adventofcode.com/2020/day/24)

import sys
from collections import defaultdict
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()


# part 1


def walk(data, tiles):
    x, y = 0, 0
    i = 0
    while i < len(data):
        if data[i : i + 2] == "ne":
            i += 2
            x = x + 1
            y = y + 1

        elif data[i : i + 2] == "se":
            i += 2
            x = x + 1
            y = y - 1

        elif data[i : i + 1] == "w":
            i += 1
            x = x - 2

        elif data[i : i + 1] == "e":
            i += 1
            x = x + 2

        elif data[i : i + 2] == "nw":
            i += 2
            x = x - 1
            y = y + 1

        elif data[i : i + 2] == "sw":
            i += 2
            x = x - 1
            y = y - 1

        else:
            raise ValueError(data[i:])

    tiles[(x, y)] = not tiles[(x, y)]


tiles = defaultdict(lambda: False)
for line in lines:
    walk(line, tiles)

# get only the black tiles (=True, white tiles=False)
# we need them for part 2
tiles = set((x, y) for (x, y), face in tiles.items() if face is True)

print(len(tiles))


# part 2

# the six neighbors of a tile
NEIGHBORS = ((1, 1), (1, -1), (-2, 0), (2, 0), (-1, 1), (-1, -1))

# Hexagonal Conway's game of life

for day in range(100):
    new_tiles = set()
    seen = set()

    # for every visited tiles
    for black_x, black_y in tiles:
        # test the tile and its six neighbors
        for dx, dy in ((0, 0),) + NEIGHBORS:
            x, y = black_x + dx, black_y + dy

            # but only once
            if (x, y) in seen:
                continue
            seen.add((x, y))

            is_black = (x, y) in tiles

            black_neighbors = sum(1 for nx, ny in NEIGHBORS if (x + nx, y + ny) in tiles)

            if is_black and (black_neighbors == 0 or black_neighbors > 2):
                new_is_black = False
            elif not is_black and (black_neighbors == 2):
                new_is_black = True
            else:
                new_is_black = is_black

            if new_is_black:
                new_tiles.add((x, y))

    tiles = new_tiles

print(len(tiles))

File: src/year2020/day24/day24.rs
//! [Day 24: Lobby Layout](https://adventofcode.com/2020/day/24)

use aoc::Coord;
use rustc_hash::{FxHashMap, FxHashSet};

struct Walk(FxHashMap<Coord, bool>);

impl Walk {
    fn new() -> Self {
        Self(FxHashMap::default())
    }

    fn tiles(&self) -> FxHashSet<Coord> {
        self.0
            .iter()
            .filter(|(_, v)| **v)
            .map(|(k, _)| k)
            .copied()
            .collect()
    }

    fn walk(&mut self, line: &str) {
        let mut pos = Coord::ZERO;

        let mut it = line.chars();
        while let Some(ch) = it.next() {
            pos += match ch {
                'w' => Coord::WEST * 2,
                'e' => Coord::EAST * 2,
                'n' => match it.next().unwrap() {
                    'w' => Coord::NORTH_WEST,
                    'e' => Coord::NORTH_EAST,
                    _ => panic!(),
                },
                's' => match it.next().unwrap() {
                    'w' => Coord::SOUTH_WEST,
                    'e' => Coord::SOUTH_EAST,
                    _ => panic!(),
                },
                _ => break,
            };
        }

        *self.0.entry(pos).or_default() ^= true;
    }
}

#[derive(Debug)]
struct Puzzle {
    tiles: FxHashSet<Coord>,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        let mut walk = Walk::new();

        for line in data.lines() {
            walk.walk(line);
        }

        Self {
            tiles: walk.tiles(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        self.tiles.len()
    }

    /// Solve part two.
    fn part2(&self) -> usize {
        let mut tiles = self.tiles.clone();

        let neighbors = [
            Coord::ZERO,
            Coord::NORTH_EAST,
            Coord::EAST * 2,
            Coord::SOUTH_EAST,
            Coord::NORTH_WEST,
            Coord::WEST * 2,
            Coord::SOUTH_WEST,
        ];

        // Hexagonal Conway's game of life

        for _ in 0..100 {
            let mut new_tiles = FxHashSet::default();
            let mut seen = FxHashSet::default();

            // for every visited tiles
            for &black in &tiles {
                for dxy in neighbors {
                    let xy = black + dxy;

                    // but only once
                    if !seen.insert(xy) {
                        continue;
                    }

                    let is_black = tiles.contains(&xy);

                    let black_neighbors = neighbors[1..]
                        .iter()
                        .filter(|&&nxy| tiles.contains(&(xy + nxy)))
                        .count();

                    // evolve rule
                    let new_is_black = if is_black && (black_neighbors == 0 || black_neighbors > 2)
                    {
                        false
                    } else if !is_black && black_neighbors == 2 {
                        true
                    } else {
                        is_black
                    };

                    if new_is_black {
                        new_tiles.insert(xy);
                    }
                }
            }

            tiles = new_tiles;
        }

        tiles.len()
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (usize, usize) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn part1() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 10);
    }

    #[test]
    fn part2() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 2208);
    }
}
